proposals/0000-grand-class-unification.rst

```diff
@@ -0,0 +1,167 @@
+The Grand Plan for Unification of Class Methods and Associated Types
+====================================================================
+
+.. proposal-number::
+.. ticket-url::
+.. implemented::
+.. highlight::
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/236>`_.
+.. sectnum::
+.. contents::
+
+NB. This is a "big picture" proposal, prompted by several other threads on the
+GHC Proposals platform. It is not intended to be submitted to the committee in
+its current form; rather, it outlines a general direction of travel for other,
+less ambitious proposals, to take into account.
+
+The proposed features are:
+
+* top-level signatures for class methods
+* remove the ``type`` prefix of top-level kind signatures
+* visible dependent quantification in terms
+* visible quantification of ambiguous class variables
+* constrained type families
+
+In more detail:
+
+* Add top-level signatures for class methods that serve as machine-checked
+  documentation::
+
+    class Eq a where
+      eq :: a -> a -> a
+
+    eq :: forall t. Eq t => t -> t -> t
+
+  The rule for checking the top-level signature is as follows. Let us refer to
+  the in-class method as ``$eq``, then we assume the following definition::
+
+    eq = $eq
+
+  This allows us to reorder and rename invisible type variables, and to
+  use a subclass instead of the actual parent class.
+
+* Remove the ``type`` prefix of top-level kind signatures::
+
+    Functor :: (Type -> Type) -> Constraint
+    class Functor f where
+      fmap :: (a -> b) -> f a -> f b
+
+  This unifies TLKS and regular, term-level signatures. In case of namespace
+  ambiguity, reject::
+
+    data T = T
+    T :: ...   -- not allowed.
+
+* Allow visible dependent quantification in terms::
```

Comment:
Again, I think there's at least one previous GHC proposal that has discussed allowing visible dependent quantification syntax in terms, so mention that.

Reply:
Is there? I couldn't find it.

Reply:
There's #102, and possibly others.